File: coding-challenges/hacker-rank/algorithms/PalindromeIndex.js
/**
 * Palindrome Index (JavaScript)
 * HackerRank Algorithm Challenge
 * https://www.hackerrank.com/challenges/palindrome-index
 *
 * michael@softwareontheshore.com
 *
 * NOTE: Currently passes half of test cases. Times out on large cases
 */
'use strict';

const testA = 'aaab';
const testB = 'baa';
const testC = 'aaa';
const testD = 'hgygsvlfcwnswtuhmyaljkqlqjjqlqkjlaymhutwsnwcwflvsgygh';
const testQ = 'abcdefg';


/* imported palindrome checker from previous challenge */
function findPalindrome(str) {
  const firstCharacter = str.slice(0, 1);
  const lastCharacter = str.slice(-1);
  const middleCharacters = str.slice(1, -1);

  if (str.length <= 1) {
    return true;
  }
  if (firstCharacter !== lastCharacter) {
    return false;
  }
  return findPalindrome(middleCharacters);
}

function findPalindromeIndex(input) {
  if (findPalindrome(input)) return -1; // if already a palindrome

  for (let x = 0; x < input.length; x++) {
    let characters = input.slice().trim().split('');
    characters.splice(x, 1);
    const text = characters.join('');

    if (findPalindrome(text)) return x;
  }
}

const result = findPalindromeIndex(testD);
console.log(result);

File: hacker-rank/algorithms/PalindromeIndex.py
"""
 * Palindrome Index (Python)
 * HackerRank Algorithm Challenge
 * https://www.hackerrank.com/challenges/palindrome-index
 *
 * NOTE: Currently passes half of test cases. Times out on large cases
"""
testA = 'aaab';
testB = 'baa';
testC = 'aaa';
testD = 'hgygsvlfcwnswtuhmyaljkqlqjjqlqkjlaymhutwsnwcwflvsgygh';
testQ = 'abcdefg';

def checkPalindrome(str):
    for x in xrange(0, len(str)):
        if str[x] != str[(len(str) - 1) - x] and x != (len(str) / 2) - 1: return False

    return True;

def findPalindromeIndex(input):
    if checkPalindrome(input): return -1

    for x in xrange(0, len(input)):
        text = input[:]
        text = text[:x] + text[x + 1:]

        if checkPalindrome(text): return x

result = findPalindromeIndex(testD)
print result
